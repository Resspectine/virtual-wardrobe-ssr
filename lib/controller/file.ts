export const fileFetch: (file: Blob, url?: string) => Promise<Response> = async (file, url = 'files/upload') => {
  const formData = new FormData();

  formData.append('file', file);

  const request = await fetch(`http://localhost:3030/api/${url}`, {
    method: 'POST',
    body: formData,
    credentials: 'include',
  });

  if (request.status === 401) {
    throw new Error('Unauthorized');
  }

  return request;
};
