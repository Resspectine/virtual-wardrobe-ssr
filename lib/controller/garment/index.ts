import { appFetch } from '..';

import { CreateRequestGarment, LoadGarmentsQueryParams, UpdateRequestGarment } from './types';
import { validateGarment } from './validations';

import { createQueryParams } from '@/lib/helpers/queryParams';
import { Garment } from 'types/garment';

export const loadGarments = async (params: LoadGarmentsQueryParams): Promise<Garment[]> => {
  const queryParams = createQueryParams(params);

  return appFetch(`garment?${queryParams}`, {
    method: 'GET',
  });
};

export const wearGarment = (garmentId: string): Promise<Response> =>
  appFetch(`garment/wear/${garmentId}`, {
    method: 'PATCH',
  });

export const triggerFavorite = async (garmentId: string): Promise<void> =>
  appFetch(`garment/favorite/${garmentId}`, {
    method: 'PATCH',
  });

export const removeGarment = async (garmentId: string): Promise<void> =>
  appFetch(`garment/${garmentId}`, {
    method: 'DELETE',
  });

export const createGarment = async (garment: CreateRequestGarment): Promise<void> => {
  if (!validateGarment(garment)) {
    return new Promise((_, reject) => {
      reject();
    });
  }

  return appFetch('garment', {
    method: 'POST',
    body: garment,
  });
};

export const getGarmentById = async (id: string | undefined): Promise<Garment | void> => {
  if (!id) {
    return undefined;
  }

  return appFetch(`garment/${id}`, {
    method: 'GET',
  });
};

export const editGarment = async (garment: Partial<UpdateRequestGarment>): Promise<void> => {
  await appFetch(`garment/${garment.id}`, {
    body: garment,
    method: 'PATCH',
  });
};
