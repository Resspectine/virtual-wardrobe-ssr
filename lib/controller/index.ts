export const appFetch: (
  input: RequestInfo,
  init?: (Omit<RequestInit, 'body'> & { body?: Record<string, unknown> }) | undefined,
  returnBlob?: boolean
) => Promise<any> = async (input, init, returnBlob) => {
  const headers = new Headers(init?.headers);

  headers.append('Content-Type', 'application/json');

  const request = await fetch(`http://localhost:3030/api/${input}`, {
    ...init,
    body: JSON.stringify(init?.body || undefined),
    headers,
    credentials: 'include',
  });

  if (request.status === 500) {
    throw new Error('Server error');
  }

  if (!returnBlob) {
    const body = await request.json();

    if (body.statusCode === 401 || body.statusCode === 400) {
      throw new Error(body.message);
    }

    return body;
  }

  return request.blob();
};
