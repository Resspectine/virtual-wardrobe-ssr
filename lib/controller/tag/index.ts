import { appFetch } from '..';

import { Tag } from '@/types/tag';

export const loadTags = async (): Promise<Tag[]> =>
  appFetch('tag', {
    method: 'GET',
  });

export const removeTag = async (tagId: string): Promise<Tag[]> =>
  appFetch(`tag/${tagId}`, {
    method: 'DELETE',
  });

export const createNewTag = async (tag: Omit<Tag, 'id'>): Promise<void> =>
  appFetch('tag', {
    method: 'POST',
    body: tag,
  });
