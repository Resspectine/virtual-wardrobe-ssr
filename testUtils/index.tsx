import { queries, Queries, render as reactRender, RenderOptions, RenderResult } from '@testing-library/react';
import { FC, ReactElement } from 'react';
import { QueryClient, QueryClientProvider } from 'react-query';

import { ThemeProvider } from '@/materialUiTheme/provider';

export const sleep = (timeout: number): Promise<void> =>
  new Promise(resolve => {
    setTimeout(() => {
      resolve();
    }, timeout);
  });

export const render = <Q extends Queries = typeof queries, Container extends Element | DocumentFragment = HTMLElement>(
  ui: ReactElement,
  options: RenderOptions<Q, Container> = {}
): RenderResult<Q, Container> => {
  const queryClient = new QueryClient();

  const Wrapper: FC = ({ children }) => (
    <QueryClientProvider client={queryClient}>
      <ThemeProvider>{children}</ThemeProvider>
    </QueryClientProvider>
  );

  return { ...reactRender(ui, { ...options, wrapper: Wrapper }) };
};
