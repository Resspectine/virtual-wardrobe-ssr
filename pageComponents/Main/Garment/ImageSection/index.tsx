import { Box } from '@mui/system';
import { FC } from 'react';

import { useImageSection } from './hooks';
import { GarmentImage, GarmentImageDivider } from './styled';

import { File as PublicFile } from '@/types/file';

export interface ImageSectionProps {
  image: PublicFile | null;
}

export const ImageSection: FC<ImageSectionProps> = ({ image }) => {
  const { imageUrl } = useImageSection({ image });

  return (
    (imageUrl && (
      <>
        <Box>
          <GarmentImage component="img" src={imageUrl} alt="Something went wrong" />
        </Box>
        <GarmentImageDivider />
      </>
    )) ||
    null
  );
};
