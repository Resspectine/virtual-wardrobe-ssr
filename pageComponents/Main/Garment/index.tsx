import Box from '@mui/material/Box';
import { FC } from 'react';

import { StarButton, DataSection, PriceSection } from './Components';
import { ImageSection } from './ImageSection';
import { GarmentWrapper } from './styled';

import TagItem from '@/components/TagItem';
import { HoldClickProps } from '@/lib/hooks/useHoldClick';
import { Garment } from '@/types/garment';

export interface GarmentProps extends Partial<HoldClickProps> {
  garment: Garment;
  onClick: () => void;
  toggleFavorite: (id: string) => void;
}

const GarmentComponent: FC<GarmentProps> = ({
  garment: { description, picture, isFavorite, price, title, wearingAmount, id, tags },
  onClick,
  toggleFavorite,
  ...hold
}) => (
  <GarmentWrapper data-testid="garment" onClick={onClick} {...hold}>
    <StarButton id={id} isFavorite={isFavorite} toggleFavorite={toggleFavorite} />
    <Box>
      {picture && <ImageSection image={picture} />}
      <DataSection description={description} title={title} />
      <Box display="flex" mx={-0.625} flexWrap="wrap" flex="1" alignItems="flex-start">
        {tags?.map(tag => (
          <TagItem tag={tag} key={tag.id} />
        ))}
      </Box>
    </Box>
    <PriceSection price={price} wearingAmount={wearingAmount} />
  </GarmentWrapper>
);

export default GarmentComponent;
