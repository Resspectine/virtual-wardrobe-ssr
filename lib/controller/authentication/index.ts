import { LoginUser, RegisterUser } from './types';
import { validateLoginUser, validateRegisterUser } from './validation';

import { appFetch } from '@/lib/controller';
import { StoreUser } from '@/store/user';

export const authenticate = async (): Promise<StoreUser | null> =>
  appFetch('authentication', {
    method: 'GET',
  });

export const login = async (user: LoginUser): Promise<StoreUser> => {
  if (!validateLoginUser(user)) {
    return Promise.reject(Error('Invalid values'));
  }

  return appFetch('authentication/log-in', {
    method: 'POST',
    body: user,
  });
};

export const register = async (user: RegisterUser): Promise<void> => {
  if (!validateRegisterUser(user)) {
    return Promise.reject(Error('Invalid values'));
  }

  return appFetch('authentication/register', {
    method: 'POST',
    body: user,
  });
};
