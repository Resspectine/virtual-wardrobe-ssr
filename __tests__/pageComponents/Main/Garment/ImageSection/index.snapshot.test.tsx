import { waitFor } from '@testing-library/react';

import { appFetch } from '@/lib/controller';
import { ImageSection } from '@/pageComponents/Main/Garment/ImageSection';
import { render } from '@/testUtils';

jest.mock('@/lib/controller');

const mockedAppFetch = appFetch as unknown as jest.Mock<ReturnType<typeof appFetch>>;

afterEach(() => {
  mockedAppFetch.mockReset();
});

afterAll(() => {
  mockedAppFetch.mockRestore();
});

describe('ImageSection', () => {
  test('should match snapshot', async () => {
    mockedAppFetch.mockResolvedValue(new Blob());

    const { getByAltText, asFragment } = render(
      <ImageSection
        image={{
          filename: '',
          id: '',
          mimetype: '',
          path: '',
        }}
      />
    );

    await waitFor(() => expect(getByAltText('Something went wrong')).toBeInTheDocument());

    expect(mockedAppFetch).toBeCalledTimes(1);

    expect(asFragment()).toMatchSnapshot();
  });
});
