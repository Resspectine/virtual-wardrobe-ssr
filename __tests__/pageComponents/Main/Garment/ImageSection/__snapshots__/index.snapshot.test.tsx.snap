// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageSection should match snapshot 1`] = `
<DocumentFragment>
  .emotion-1 {
  width: 100%;
  height: auto;
}

<div
    class="MuiBox-root emotion-0"
  >
    <img
      alt="Something went wrong"
      class="MuiBox-root emotion-1"
      src="data:application/octet-stream;base64,"
    />
  </div>
  .emotion-0 {
  height: 1px;
  width: 90%;
  border-radius: 1px;
  background-color: #e8e8e8;
  margin: 10px auto;
}

<div
    class="MuiBox-root emotion-0"
  />
</DocumentFragment>
`;
